app/src/main/java/io/appium/settings/RecorderThread.java

```diff
@@ -0,0 +1,419 @@
+            super.onStopped();
+            if (!stopped) {
+                asyncError = true;
+            }
+        }
+    };
+
+    public RecorderThread(MediaProjection mediaProjection, String outputFilePath,
+                          int videoWidth, int videoHeight, boolean isRotated) {
+        this.mediaProjection = mediaProjection;
+        handler = new Handler();
+        this.outputFilePath = outputFilePath;
+        this.videoWidth = videoWidth;
+        this.videoHeight = videoHeight;
+        this.sampleRate = RecorderConstant.AUDIO_CODEC_SAMPLE_RATE;
+        this.isRotated = isRotated;
+        videoMime = MediaFormat.MIMETYPE_VIDEO_AVC;
+        bufferInfo = new MediaCodec.BufferInfo();
+        lastAudioTimestampUs = -1;
+    }
+
+    public void startRecording() {
+        stopped = false;
+        recordingThread = new Thread(this);
+        recordingThread.start();
+    }
+
+    public void stopRecording() {
+        stopped = true;
+    }
+
+    public boolean isRecordingContinue() {
+        return stopped;
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private void setupVideoCodec() throws IOException {
+        // Encoded video resolution MUST match virtual display.
+        videoEncoder = MediaCodec.createEncoderByType(videoMime);
+
+        /* Clamp raw width/height of device screen with video encoder's capabilities
+        *  to avoid crash.
+        * */
+        MediaCodecInfo.VideoCapabilities videoEncoderCapabilities = videoEncoder.getCodecInfo().
+                getCapabilitiesForType(videoMime).getVideoCapabilities();
+        this.videoWidth = videoEncoderCapabilities.getSupportedWidths().clamp(this.videoWidth);
+        this.videoHeight = videoEncoderCapabilities.getSupportedHeights().clamp(this.videoHeight);
+        this.videoBitrate = calcBitRate(videoWidth, videoHeight);
+
+        MediaFormat encoderFormat = MediaFormat.createVideoFormat(videoMime, videoWidth,
+                videoHeight);
+        encoderFormat.setInteger(MediaFormat.KEY_COLOR_FORMAT,
+                MediaCodecInfo.CodecCapabilities.COLOR_FormatSurface);
+        encoderFormat.setInteger(MediaFormat.KEY_BIT_RATE, videoBitrate);
+        encoderFormat.setInteger(MediaFormat.KEY_FRAME_RATE,
+                RecorderConstant.VIDEO_CODEC_FRAME_RATE);
+        encoderFormat.setInteger(MediaFormat.KEY_REPEAT_PREVIOUS_FRAME_AFTER,
+                RecorderConstant.AUDIO_CODEC_REPEAT_PREV_FRAME_AFTER);
+        encoderFormat.setInteger(MediaFormat.KEY_I_FRAME_INTERVAL,
+                RecorderConstant.AUDIO_CODEC_I_FRAME_INTERVAL);
+
+        videoEncoder.configure(encoderFormat, null, null,
+                MediaCodec.CONFIGURE_FLAG_ENCODE);
+        surface = videoEncoder.createInputSurface();
+        videoEncoder.start();
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private void setupVirtualDisplay() {
+        virtualDisplay = mediaProjection.createVirtualDisplay("Appium Screen Recorder",
```

Comment:
consider this method to return VirtualDisplay instance rather that being void. Having so many depending methods creates a dependency clear where it is not were clear in which order they must be called and when.

Reply:
done

Reply:
other `setupXX` methods also starts instance inside them (e.g `setupAudioEncoder` -> `audioEncoder.start()`) should i change them too? (i will put `.start()` calls to outside of function)

Reply:
up to you. I've wrote an extended comment above regarding the common variable scoping strategy